index.js

```diff
@@ -8,32 +8,48 @@ var Reader = module.exports = function(options) {
   options = options || {}
   this.offset = 0
   this.lastChunk = false
-  this.chunk = null
+  this.chunk = Buffer.alloc(4);
+  this.chunkLength = 0;
   this.headerSize = options.headerSize || 0
   this.lengthPadding = options.lengthPadding || 0
   this.header = null
   assert(this.headerSize < 2, 'pre-length header of more than 1 byte length not currently supported')
 }
 
 Reader.prototype.addChunk = function(chunk) {
-  this.offset = 0
-  this.chunk = chunk
-  if(this.lastChunk) {
-    this.chunk = Buffer.concat([this.lastChunk, this.chunk])
-    this.lastChunk = false
+  var newChunkLength = chunk.length;
+  var newLength = this.chunkLength + newChunkLength;
+
+  if (newLength > this.chunk.length) {
+    var newBufferLength = this.chunk.length * 2;
+    while (newLength >= newBufferLength) {
+      newBufferLength *= 2;
+    }
+    var newBuffer = new Buffer(newBufferLength);
+    this.chunk.copy(newBuffer);
+    this.chunk = newBuffer;
+  }
+  chunk.copy(this.chunk, this.chunkLength);
+  this.chunkLength = newLength;
+
+  // If more than half of the data has been read, shrink
+  // the buffer and reset the offset to reclaim the memory
+  var halfLength = this.chunk.length / 2;
+  if (this.offset > halfLength) {
+    var newBuffer = new Buffer(halfLength);
+    this.chunk.copy(newBuffer, 0, this.offset);
+    this.chunk = newBuffer;
+    this.chunkLength -= this.offset;
+    this.offset = 0;
   }
 }
 
 Reader.prototype._save = function() {
```

Comment:
yeah if this method is only returning `false` and doing nothing else we might as well inline it (though v8 probably is anyway)